### PR TITLE
Fix int4 KV cache quantization range when the packed head width is not a power of two

### LINKED ISSUE #4849: [Bug] int4 KV cache: the quantization range is stretched to include 0 when the packed head width is not a power of two
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.

### Describe the bug

`_fill_page_quant_int4` tiles the head at `BLOCK_D = next_power_of_2(head_dim)`, where `head_dim` is the packed (half) width. When that width is not a power of two the extra lanes fall past the end of the head, and because they are masked at the load they arrive as `0.0` and still take part in the min/max reduction in `_quant_int4`. The range becomes `[min(0, lo), max(0, hi)]` instead of `[lo, hi]`, so every value in that head is quantized on a range wider than the data occupies. A raw `head_dim` of 96 packs to 48 and gives `BLOCK_D` 64 — a quarter of every tile is dead lanes.

### Reproduction

```python
import torch

from lmdeploy.messages import QuantPolicy
from lmdeploy.pytorch.kernels.cuda.fill_kv_cache import fill_kv_cache

HEAD_DIM, NUM_HEADS, NUM_TOKENS, BLOCK_SIZE = 96, 2, 4, 16  # packed width 48 -> BLOCK_D 64, 16 dead lanes


def quant(kv, nbits=4):  # verbatim from tests/pytorch/kernel/test_fill_kv_cache.py
    amax, amin = kv.amax(dim=-1, keepdim=True), kv.amin(dim=-1, keepdim=True)
    scales = (amax - amin) / (2**nbits - 1)
    zeros = -amin / scales
    q = (kv / scales + zeros + 0.5).to(torch.uint8)
    q1, q2 = q.split(q.shape[-1] // 2, -1)
    return q1 + q2 * 16, torch.cat([scales, zeros], dim=-1)


torch.manual_seed(0)
# a head whose values do not straddle zero
states = torch.rand((NUM_TOKENS, NUM_HEADS, HEAD_DIM), device='cuda') * 2 + 1
k_caches = torch.zeros((1, BLOCK_SIZE, NUM_HEADS, HEAD_DIM // 2), dtype=torch.uint8, device='cuda')
v_caches, k_sz = torch.zeros_like(k_caches), torch.zeros((1, BLOCK_SIZE, NUM_HEADS, 2), device='cuda')
v_sz = torch.zeros_like(k_sz)

fill_kv_cache(states, states, k_caches, v_caches,
              torch.tensor([0], device='cuda'), torch.tensor([NUM_TOKENS], device='cuda'),
              torch.tensor([NUM_TOKENS], device='cuda'), NUM_TOKENS, torch.tensor([[0]], device='cuda'),
              k_sz, v_sz, QuantPolicy.INT4)

ref_q, ref_sz = quant(states)
got_q, got_sz = k_caches[0, :NUM_TOKENS], k_sz[0, :NUM_TOKENS]
print(f'values in the head: [{states.min():.4f}, {states.max():.4f}]')
print(f'scale   kernel={got_sz[0, 0, 0]:.6f}  expected={ref_sz[0, 0, 0]:.6f}')
print(f'zero    kernel={got_sz[0, 0, 1]:.6f}  expected={ref_sz[0, 0, 1]:.6f}')
print(f'packed bytes differing from the reference: {int((got_q != ref_q).sum())}/{ref_q.numel()}')


def dequant(q, sz):
    full = torch.cat([(q & 0xF).float(), (q >> 4).float()], dim=-1)
    return (full - sz[..., 1:2]) * sz[..., 0:1]


print(f'max dequant error   kernel={(dequant(got_q, got_sz) - states).abs().max():.6f}  '
      f'expected={(dequant(ref_q, ref_sz) - states).abs().max():.6f}')
```

```
values in the head: [1.0025, 2.9987]
scale   kernel=0.199023  expected=0.131584
zero    kernel=0.000000  expected=-7.687806
packed bytes differing from the reference: 379/384
max dequant error   kernel=0.099709  expected=0.065959
```

The step is 51% larger than it needs to be, i.e. about 0.6 of the 4 bits is spent covering a part of the range the data never reaches. With `head_dim=128` (packed 64), or with values that straddle zero, the kernel matches the reference exactly.

### Environment

```Shell
lmdeploy: main @ 98058c354a83f25f52b5d5ea80bfb4686996451e, installed with DISABLE_TURBOMIND=1 pip install -e .
GPU: NVIDIA B200, driver 595.71.05 (CUDA 13.2)
torch 2.12.1+cu130, triton 3.7.1, Python 3.11
```

### PR DESCRIPTION
**Pull request by truong-v**
## Motivation

Fixes #4849.

## Modification

- `_fill_page_quant_int4`: add `d_off = d_off % head_dim` so the lanes past the head wrap onto real channels, and load `state0`/`state1` with `kv_mask[:, None]` instead of `mask_kc` so those lanes carry data rather than `0.0`. This is exactly the shape `_fill_page_quant_int8` has. The store keeps `mask_kc`, so nothing extra is written, and every address stays inside the head.
- `TestFillKVCacheInt4NonPow2`: a subclass of the existing int4 test at `head_dim` 96 and 128, with states drawn away from zero. Both are needed — 128 packs to a power of two, and states that straddle zero make `[min(0, lo), max(0, hi)]` coincide with `[lo, hi]`, which is why the current int4 test (`head_dim` 128 only, `randn` states) cannot see this. Reverting the kernel change alone fails the two `head_dim=96` cases and leaves everything else green.

Addressing-only change, so it is perf neutral; `fill_kv_cache` int4 prefill, three interleaved before/after runs on a B200:

| shape | before | after |
|---|---|---|
| head_dim 96, 8 heads, 4096 tokens | 12.30-12.31 us | 12.31-12.33 us |
| head_dim 128, 8 heads, 4096 tokens | 12.55-12.62 us | 12.56-12.57 us |
| head_dim 96, 32 heads, 8192 tokens | 38.86-38.90 us | 38.91-38.92 us |
| head_dim 128, 32 heads, 8192 tokens | 39.02-39.05 us | 39.02-39.04 us |

## Environment

```
lmdeploy: main @ 98058c354a83f25f52b5d5ea80bfb4686996451e, installed with DISABLE_TURBOMIND=1 pip install -e .
GPU: NVIDIA B200, driver 595.71.05 (CUDA 13.2)
torch 2.12.1+cu130, triton 3.7.1, Python 3.11
```

### CODE REVIEW

**Review by grimoire — APPROVED** [review]
LGTM
